Improve error handling of `x509_v3_set_ctx` and `x509_v3_set_nconf`

DDvO opened the pull request:
These two commits:

* `X509V3_set_nconf()`: Improve error handling using this function, mostly in `apps/`
* `X509V3_set_ctx()`: Improve error handling and documentation

have been carved out from #16342 to simplify the latter. 
They bring minor improvements to the mentioned lib functions and their use in `apps/`, 
where two further auxiliary functions are added to `apps.c`.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated
- [ ] tests are added or updated

DDvO commented:
Does this make sense also for openssl-3.0?

t8m commented:
> Does this make sense also for openssl-3.0?

I am not sure this will qualify as pure bug fix so IMO no.

t8m commented on apps/req.c:
```diff
@@ -558,24 +558,20 @@ int req_main(int argc, char **argv)
-        if (!X509V3_EXT_add_nconf(addext_conf, &ctx, "default", NULL)) {
-            BIO_printf(bio_err, "Error checking extensions defined using -addext\n");
+        if (!do_EXT_add_nconf(addext_conf, &ctx, "default", NULL,
+                              "Error checking x509 extensions defined via -addext%.0s\n"))
```
This is really ugly. Could you just make the printf in do_EXT_add_nconf() to not add the sect if it is NULL? But then it risks problems if sect value can be NULL in other cases... Perhaps add another parameter?

DDvO replied:
All right, changed this to allow a NULL `sect` arg.
I think for such an app-internal function it is bearable that the `msg` arg must contain `%s` if and only if the `sect` arg is non-NULL.

t8m commented on include/openssl/x509v3.h.in:
```diff
@@ -649,16 +649,16 @@ int X509V3_add_value_bool_nf(const char *name, int asn1_bool,
-void X509V3_set_nconf(X509V3_CTX *ctx, CONF *conf);
+int X509V3_set_nconf(X509V3_CTX *ctx, CONF *conf);
 void X509V3_set_conf_lhash(X509V3_CTX *ctx, LHASH_OF(CONF_VALUE) *lhash);
 # endif
 
 char *X509V3_get_string(X509V3_CTX *ctx, const char *name, const char *section);
 STACK_OF(CONF_VALUE) *X509V3_get_section(X509V3_CTX *ctx, const char *section);
 void X509V3_string_free(X509V3_CTX *ctx, char *str);
 void X509V3_section_free(X509V3_CTX *ctx, STACK_OF(CONF_VALUE) *section);
-void X509V3_set_ctx(X509V3_CTX *ctx, X509 *issuer, X509 *subject,
-                    X509_REQ *req, X509_CRL *crl, int flags);
+int X509V3_set_ctx(X509V3_CTX *ctx, X509 *issuer, X509 *subject,
```
Now, the question is whether changing return value from void to int is an acceptable change for a minor release.

DDvO replied:
From the type safety perspective it is not a problem to add a return type (or to make the return type more specific in case there was already a return type).

t8m replied:
IMO OTC will have to decide.

t8m commented:
> I am not sure this will qualify as pure bug fix so IMO no.

And to make it even stronger changing the return value of functions from void to int is something that is definitely not acceptable for a stable release branch.

t8m commented on apps/req.c:
```diff
@@ -844,21 +841,16 @@ int req_main(int argc, char **argv)
-                                         new_x509)) {
-                BIO_printf(bio_err, "Error adding x509 extensions defined via -addext\n");
+                && !do_EXT_add_nconf(addext_conf, &ext_ctx, "default", new_x509,
+                                     "Error adding x590 extensions defined via -addext%0s\n"))
```
You have %0s here as well.

DDvO replied:
Oops, fixed, as well as a typo in this message string (x590 vs. x509).

DDvO commented:
BTW, I added today further clarifications to the documentation and the code regarding the use of the`subject` and `req` parameters of `X509V3_set_ctx`.

Also documented the `email:move` config option value, which so far was missing from `x509v3_config.pod`.

DDvO commented:
>>  Does this make sense also for openssl-3.0?
> 
> I am not sure this will qualify as pure bug fix so IMO no.

> And to make it even stronger changing the return value of functions from void to int is something that is definitely not acceptable for a stable release branch.

>>> Now, the question is whether changing return value from void to int is an acceptable change for a minor release.
>> 
>> From the type safety perspective it is not a problem to add a return type (or to make the return type more specific in case there was already a return type).
> 
> IMO OTC will have to decide.

Since the error handling additions to `x509_v3_set_ctx()` and `x509_v3_set_nconf()` are on the one hand a (minor) bug fix but on the other hand require adding a return value, which according to the quoted answers cannot be done for openssl-3.0, it seems this PR can only be merged for `master`. 

Does inclusion in 3.1(-dev) require an OTC decision?
If so, how about approving this now while adding the labels **master** and **hold: need OTC decision**?

t8m commented:
**Summary of what are the things to decide here by OTC:**

This PR changes return value of public API calls `X509V3_set_nconf()` and `X509V3_set_ctx()` from `void` to `int`. This is necessary (unless we want to add _ex calls with the only difference being the return type) to be able to properly report errors from these calls. (This PR adds that reporting in apps, which is the uncontroversial part of the PR.)

The question is - is such change acceptable for the next minor release or not?
(I do not expect such change being acceptable for 3.0.)

paulidale commented:
OTC: such a change would be for a major release.

DDvO commented:
Thanks @t8m and @paulidale for the clarification when this PR can be merged.
Consequently, this PR will be on hold for quite a while.
So I carved out two rather independent non-critical commits to #17145.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

DDvO commented:
Carved out the general doc improvements to #17358 such that they are made available earlier and potential further merge conflicts are avoided for them.

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 61 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 92 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 123 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 154 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 185 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 216 days ago

DDvO commented:
How best to deal for now with PRs that are targeted for 4.0?
Remove the 'pending' flags to avoid reminders to pop up regularly?

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 92 days ago

DDvO commented:
>Nice improvements.

Thanks @paulidale for your approval and comment.

Asking @openssl/committers for 2nd review.

t8m reviewed (changes requested):
One doc nit.

DDvO commented:
@t8m @paulidale since I fixed the doc nit, please don't forget to (re-)approve such that we can get this finished off.

t-j-h commented on crypto/x509/v3_conf.c:
```diff
@@ -444,31 +444,39 @@ static X509V3_CONF_METHOD nconf_method = {
     ctx->issuer_pkey = NULL;
+
+    /* doing this check after initialization as a precaution */
+    if ((subject != NULL) + (req != NULL) + (crl != NULL) > 1) {
```
It wouldn't hurt to actually add a comment there that at most one of these three parameters can be non-null - then the reader will get the nature of the check.

And although the doc currently says this is was **never** previously enforced so this is actually a behaviour change - remembering that the code trumps documentation, especially since that documentation dates from 2021.

**And there is definitely code out there which sets multiple pointers at once.**

I don't think we should add documentation that the code does not actually enforce and then change the code to enforce that documentation - when there is nothing in API history or usage that supports that particular interpretation - and there are users of the API which clearly depend on the behaviour. It is not hard to see that code usage out there.

t-j-h commented:
Added hold given that this code is enforcing a change which wasn't previously in place and although documentation was changed back in 2021 to say this, the code has never previously had any such restrictions.

There is a very common pattern of setting subject and issuer at the same time. We can debate the merits of the API approach - but the API is what the API is. 

Any such behaviour change that will knowingly break existing code should be called out explicity so reviewers can form an opinion on the merits of the change (knowing there will be a downstream cost). 

Another fix which could be done is to remove the incorrect documentation that was added back in 2021 - documenting a behaviour that was never present in the code (until this PR).

DDvO commented:
> There is a very common pattern of setting subject and issuer at the same time. 

Please note that the new check in `X509V3_set_ctx()` does not cover the simultaneous use of `subject` and `issuer`, which indeed is fine,
and that also the documentations just talks about the simultaneous use of `subject`, `req`, and `crl`.
Moreover, the check does not really enforce that these three are used exclusively.

> Any such behaviour change that will knowingly break existing code 

This PR does not really break existing code because, as a precaution, it is done just before function exit,
and existing code does not check an error return because it assumes that the function does not return a value.

Nevertheless, for maximal backward compatibility, I've just removed the addition of the `ERR_R_PASSED_INVALID_ARGUMENT` error queue entry on failure of the check.

Moreover, I've added now an entry in `CHANGES.md` highlighting that the two functions now return a result
and that this is done in a backward compatible way.

Yet still this PR is considered an API-breaking change simply because the two functions now return a result,
which is why it had been delayed so far until it can now be included in 4.0.

> Another fix which could be done is to remove the incorrect documentation that was added back in 2021 - documenting a behaviour that was never present in the code (until this PR).

I've relaxed the documentation to reflect the weak (non-enforcing) check, changing
`At most one of these three parameters can be non-NULL.`
to
`At most one of these three parameters should be non-NULL.`

DDvO commented:
@openssl/committers this PR has been waiting for inclusion for more than 4 years.

Since the freeze for the 4.0 alpha release is going to be on Feb 24th,
looks like this is needs to be handled within the next 5 days in order to get included, right?

I answered all open comments 2 weeks and 5 days ago, respectively.
Can the "hold: discuss" now be lifted, and this finally get approved?

DDvO commented on crypto/x509/v3_conf.c:
```diff
@@ -444,31 +444,41 @@ static X509V3_CONF_METHOD nconf_method = {
     }
     ctx->db_meth = &nconf_method;
     ctx->db = conf;
+    return 1;
```
I see your points and generally would agree on them.
Would have been good to raise and discuss them much earlier, over the last 4+ year.
Yet now we are extremely close to the 4.0 deadline, 
and it's simply too late to (re-)start making changes again,
and then risk waiting for two approvals again.

DDvO commented:
Thank you @simo5 for your approval.
@paulidale could you please re-approve, and/or @t8m give your blessing now,
such that we finally get this through the door?

paulidale commented on include/openssl/x509v3.h.in:
```diff
@@ -684,15 +684,15 @@ int X509V3_add_value_bool_nf(const char *name, int asn1_bool,
 void X509V3_string_free(X509V3_CTX *ctx, char *str);
 void X509V3_section_free(X509V3_CTX *ctx, STACK_OF(CONF_VALUE) *section);
-void X509V3_set_ctx(X509V3_CTX *ctx, X509 *issuer, X509 *subject,
+int X509V3_set_ctx(X509V3_CTX *ctx, X509 *issuer, X509 *subject,
```
```suggestion
__owur int X509V3_set_ctx(X509V3_CTX *ctx, X509 *issuer, X509 *subject,
```

slontis commented:
It looks like previous discussion within OTC thought this PR was ok.

slontis commented:
@t8m are you fine with these changes? It looks like OTC previously discussed it and decided it was ok.

DDvO commented on crypto/x509/v3_conf.c:
```diff
@@ -446,24 +446,16 @@ static X509V3_CONF_METHOD nconf_method = {
-    if (ctx == NULL) {
-        ERR_raise(ERR_LIB_X509V3, ERR_R_PASSED_NULL_PARAMETER);
-        return;
-    }
```
changed
